app/services/publisher_statement_getter.rb

```diff
@@ -1,53 +1,100 @@
+    transactions.sort_by { |x| x.created_at }
   end
 
   private
 
-  def filter_transactions_by_period(transactions, period)
-    case period
-      when "all"
-        transactions
-      when "this_month"
-        cutoff = Date.today.beginning_of_month
-        transactions.select { |transaction|
-          transaction["created_at"].to_date.at_beginning_of_month == cutoff
-        }
-      when "last_month"
-        cutoff = (Date.today - 1.month).at_beginning_of_month
-        transactions.select { |transaction|
-          transaction["created_at"].to_date.at_beginning_of_month == cutoff
-        }
-      else
-        transactions
-    end
-  end
-
   def replace_account_identifiers_with_titles(transactions)
-    transactions.map { |transaction|
+    transactions.map do |transaction|
       account_identifier = transaction["channel"]
       if account_identifier.starts_with?(Publisher::OWNER_PREFIX)
-        transaction["channel"] = "All"
+        transaction["channel"] = I18n.t("publishers.statements.index.account")
       elsif account_identifier.blank?
         transaction["channel"] = "Manual"
       else
-        channel = Channel.find_by_channel_identifier(account_identifier)
-        transaction["channel"] = channel&.publication_title || account_identifier
+        transaction["channel"] = channel_name(account_identifier)
       end
-      transaction
-    }
+
+      Statement.new(
+        channel: transaction["channel"],
+        description: transaction["description"],
+        transaction_type: transaction["transaction_type"],
+        amount: transaction["amount"]&.to_d,
+        settlement_currency: transaction["settlement_currency"],
+        settlement_amount: transaction["settlement_amount"]&.to_d,
+        settlement_destination_type: transaction["settlement_destination_type"],
+        settlement_destination: transaction["settlement_destination"],
+        created_at: transaction["created_at"].to_date,
+      )
+    end
+  end
+
+  def get_uphold_transactions
+    uphold = []
+
+    publisher.uphold_connection.uphold_connection_for_channels.each do |card_connection|
+      transactions = card_connection.uphold_connection.uphold_client.transaction.all(id: card_connection.card_id)
+      next if transactions.blank?
+
+      transactions.each do |transaction|
+        uphold << Statement.new(
+          channel: card_connection.channel.details.publication_title,
+          transaction_type: Statement::UPHOLD_CONTRIBUTION,
+          amount: transaction.origin.dig("amount")&.to_d,
+          settlement_currency: transaction.destination.dig("currency"),
+          settlement_amount: transaction.destination.dig("amount")&.to_d,
+          created_at: transaction.createdAt.to_date,
+        )
+      end
+    end
+
+    # Replicate the similar behavior of Eyeshade. Aggregates all the contribution amounts for the month, and puts them into one entry
+    uphold.group_by { |u| u.created_at.at_end_of_month }.each do |date, entries|
```

Comment:
I'm very confused by this block. So we're going through a list of uphold contributions and somehow the settlement statements gets appended in this block?

Reply:
When we mark a transaction as paid out then we subtract from the Publishers current balance on Eyeshade so the transactions appear as negative. 

In order to actually get a valid grouping for periods for which months we paid out I select all the transactions that are negative, then group them by then select the transactions which are within the start and end of that grouping.

This was the easiest way for finding the ranges for when publishers got paid. So for Uphold it’s a little different. Technically we are have technically already paid these transactions out as they happen in real time. However if we did mark them as negative BAT amounts and the positive amount that the user got paid then our monthly groupings would be wrong. 

In order to keep the same groupings and also include the summary for how many Uphold tips there were I summarize them after I append them to all the transactions.